packages/block-editor/src/store/selectors.js

```diff
@@ -1806,26 +1833,50 @@ export const __experimentalGetParsedPattern = createSelector(
 	( state ) => [ state.settings.__experimentalBlockPatterns ]
 );
 
+export const __experimentalGetAvailableStageOneParsedPatterns = createSelector(
+	( state ) => {
+		const patterns = state.settings.__experimentalBlockPatterns;
+		const { allowedBlockTypes } = getSettings( state );
+		return patterns
+			.map( ( pattern ) => ( {
+				...pattern,
+				stageOneBlocks: stageOneParse( pattern.content ),
+			} ) )
+			.filter( ( { stageOneBlocks } ) => {
+				const isAllowed = checkAllowListRecursive(
+					stageOneBlocks,
+					allowedBlockTypes
+				);
+
+				return isAllowed;
+			} );
+	},
+	( state ) => [
+		state.settings.__experimentalBlockPatterns,
+		state.settings.allowedBlockTypes,
+	]
+);
+
```

Comment:
@david-szabo97: I don't know if it's residual or intentional, but I don't think we should break out the parsing into a new selector. The parsing should just happen inside `getAllowedPatterns`, which incidentally means we don't have to map data, just filter it:

```js
patterns.filter( ( pattern ) => {
  const parsedBlocks = parse( pattern.content );
  return (
      checkAllowListRecursive( parsedBlocks, allowedBlockTypes ) &&
      parsedBlocks.every( ( { blockName } ) =>
        canInsertBlockType( state, blockName, rootClientId )
  );
} )
```

My concerns are: conceptually, `__experimentalGetAvailableStageOneParsedPatterns` will on its own be a confusing selector for most people; in terms of API, it's an added burden for the consumers and for us to maintain; and it's an unnecessary performance penalty, as the least concern.

Finally, you're raising an interesting question: in Gutenberg, we haven't been good at properly distinguishing, in documentation and in-code terminology, between what I described as stage I and stage II parsing. This notion of stages comes from early discussions and presentations until @mtias and I included them in [The Language of Gutenberg](https://lamda.blog/2018/04/22/the-language-of-gutenberg/), but it's not official. The result is at times quite awkward, as can be seen in the name `stageOneParser`. :) Traditionally, over time, we have seen `parsedBlocks` as one of the names for those raw blocks (stage I), hence my suggestion in the snippet.

Reply:
> Traditionally, over time, we have seen parsedBlocks as one of the names for those raw blocks (stage I), hence my suggestion in the snippet.

Possibly a thought more for #15674 but I also had trouble with this naming, which is what led me to toss out `block object` vs. `block`. "Blocks" as we know them don't exist apart from the coupling of their JavaScript code and the data contained in the attributes, _the other thing we also call the block_.

Reply:
`__experimentalGetAvailableStageOneParsedPatterns` is an optimization here. We want to use memoization here to avoid reparsing all the patterns again. Since `getAllowedPatterns` memoizes based on `clientId`, it would reparse all the patterns for every different `clientId`.

I really had no idea how to name these functions and variables 😄 A comment would definitely help here.

Reply:
Ah, of course, I didn't think about how the first part is independent of `clientId`. That seems like something useful to cache if many patterns are registered. So in terms of naming I would focus more on the data relationships and less on the parsing. For instance, `getAllowedBlockPatterns`.

Keeping it private to the module is also a useful signal, just like `checkAllowListRecursive` — which is made more difficult by testing, I know. But it's something to weigh. :)